src/middlewared/middlewared/plugins/acme_protocol.py

```diff
@@ -0,0 +1,363 @@
+import json
+import requests
+import time
+
+from middlewared.schema import Bool, Dict, Int, Str, ValidationErrors
+from middlewared.service import accepts, CallError, CRUDService
+from middlewared.validators import validate_attributes
+
+from acme import client, messages
+from botocore import exceptions as boto_exceptions
+from botocore.errorfactory import BaseClientExceptions as boto_BaseClientException
+from certbot import achallenges
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives.asymmetric import rsa
+
+
+'''
+>>> issue_cert
+{'name': 'check_cert', 'tos': True, 'csr_id': 14, 'acme_directory_uri': 'https://acme-staging-v02.api.letsencrypt.org/directory', 'create_type': 'CERTIFICATE_CREATE_ACME', 'dns_mapping': {'acmedev.agencialivre.com.br': 1}}
+'''
+
+'''
+TODO'S:
+1) IS THERE A MIN KEY SIZE REQUIRED BY LETS ENCRYPT IN CSR - HANDLE THE EXCEPTION GRACEFULLY
+2) CHECK IF _GET_INSTANCE CAN BE CALLED FROM MIDDLEWARE.CALL_SYNC
+3) Domain names should not end in periods ? research
+4) Integrate alerts
+5) See what can be done to respect rate limits
+'''
+
+
+class ACMERegistrationService(CRUDService):
+
+    class Config:
+        datastore = 'system.acmeregistration'
+        datastore_extend = 'acme.registration.register_extend'
+        namespace = 'acme.registration'
+        private = True
+
+    async def register_extend(self, data):
+        data['body'] = {
+            key: value for key, value in
+            (await self.middleware.call(
+                'datastore.query', 'system.acmeregistrationbody',
+                [['acme', '=', data['id']]], {'get': True}
+            )).items() if key != 'acme'
+        }
+        return data
+
+    def get_directory(self, acme_directory_uri):
+        try:
+            response = requests.get(acme_directory_uri).json()
+            return messages.Directory({
+                key: response[key] for key in ['newAccount', 'newNonce', 'newOrder', 'revokeCert']
+            })
+        except (requests.ConnectionError, requests.Timeout, json.JSONDecodeError, KeyError) as e:
+            raise CallError(f'Unable to retrieve directory : {e}')
+
+    @accepts(
+        Dict(
+            'acme_registration_create',
+            Bool('tos', default=False),
+            Dict(
+                'JWK_create',
+                Int('key_size', default=2048),
+                Int('public_exponent', default=65537)
+            ),
+            Str('acme_directory_uri', required=True),
+        )
+    )
+    def do_create(self, data):
+
+        # STEPS FOR CREATION
+        # 1) CREATE KEY
+        # 2) REGISTER CLIENT
+        # 3) SAVE REGISTRATION OBJECT
+        # 4) SAVE REGISTRATION BODY
+
+        verrors = ValidationErrors()
+
+        directory = self.get_directory(data['acme_directory_uri'])
+        if not isinstance(directory, messages.Directory):
+            verrors.add(
+                'acme_registration_create.directory_uri',
```

Comment:
This should always match field name that has failed validation (in this case `acme_registration_create.acme_directory_uri`)